presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,182 @@
+
+    public static <T extends Node> ScopeAware<T> scopeAwareKey(T node, Analysis analysis, Scope scope)
+    {
+        return new ScopeAware<>(analysis, scope, node);
+    }
+
+    public T getNode()
+    {
+        return node;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return hash;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ScopeAware<T> other = (ScopeAware<T>) o;
+        checkArgument(this.scope == other.scope, "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                // For subqueries of the query associated with the current expression, compare by syntax
+                if (leftScope.hasOuterParent(scope) && rightScope.hasOuterParent(scope)) {
+                    return leftExpression.equals(rightExpression);
+                }
+
+                // Otherwise, for references that come from the current query scope or an outer scope of the current
+                // expression, compare by resolved field
+                return leftField.getFieldId().equals(rightField.getFieldId());
+            }
+        }
+
+        if (!left.shallowEquals(right)) {
+            return false;
+        }
+
+        List<? extends Node> leftChildren = left.getChildren();
+        List<? extends Node> rightChildren = right.getChildren();
+
+        if (leftChildren.size() != rightChildren.size()) {
+            return false;
+        }
+
+        for (int i = 0; i < leftChildren.size(); i++) {
+            if (!equal(leftChildren.get(i), rightChildren.get(i))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private int hash(Node node)
+    {
+        if (node instanceof Expression) {
+            Expression expression = (Expression) node;
+            if (analysis.isColumnReference(expression)) {
+                ResolvedField field = analysis.getResolvedField(expression);
+
+                Scope resolvedScope = field.getScope();
+                if (resolvedScope.hasOuterParent(scope)) {
```

Comment:
this still doesn't work.

Let's suppose `resolvedScope.hasOuterParent(scope) == true`, then we return `expression.hashCode()` here.

However, in comparison:
```
rightScope.hasOuterParent(scope) == false
``` 
then we compare by `leftField.getFieldId().equals(rightField.getFieldId());` which could return true.

However, hashes would not be equal.

Reply:
Indeed, described in isolation, it has that issue. But remember that ScopeAware requires that the expressions being compared be in the same "query scope" (i.e., the "root" scopes need to be local to each other). So, consider the following cases:

* If the expressions are simple expressions with no subqueries, all field references will have `hasOuterParent(scope) == false`, to this trivially works.
* If the expressions contain subqueries, the non-nested part has to be equal per the item above. The subqueries need to be structurally similar (nodes need to be shallowly equal) before recursing, so if both expressions recurse into the subquery scope at the same time.

This is not very obvious, and I almost forgot about it. I even had a discussion with Kasia about it a few weeks ago. I'll add a comment to explain so we don't forget in the future.